README.md
# [Flutter Material Dialogs](https://pub.dev/packages/material_dialogs)

[![pub](https://img.shields.io/pub/v/material_dialogs.svg)](https://pub.dev/packages/material_dialogs/install)

# Flutter Material Dialogs 📱

A Flutter library aims to help you create 💪🏻*animated*, 😃 _simple_, 😎 _stylish_ Material Dialogs in your app.

<table style="width:100%">
  <tr>
    <th><b>1. Material Dialog</b></th>
    <th><b>2. Animations Material Dialog</b></th> 
    <th><b>3. Bottom Material Dialog</b></th>
    <th><b>4. Animations Bottom Dialog</b></th>
  </tr>
  <tr>
    <td><img src="https://github.com/Ezaldeen99/material_dialogs/raw/master/gifs/normal.gif"/></td>
    <td><img src="https://github.com/Ezaldeen99/material_dialogs/raw/master/gifs/normal_animated.gif"/></td> 
    <td><img src="https://github.com/Ezaldeen99/material_dialogs/raw/master/gifs/bottom.gif"/></td>
    <td><img src="https://github.com/Ezaldeen99/material_dialogs/raw/master/gifs/bottom_animated.gif"/></td>
  </tr>
</table>

# Table of Contents:

> - [ Introduction ](#introduction)
> - [ Types of Dialog ](#types)
> - [ Implementation ](#implementation)
>   - [ Create Dialog Instance ](#createDialogInstance)
>     - [ Material Dialog ](#createMaterialDialog)
>     - [ Bottom Sheet Material Dialog ](#createBsMaterialDialog)
>   - [ Show Animations ](#showAnims)
>   - [Icon Buttons](#icons_buttons)
> - [ Contribute ](#contribute)
> - [ Credits ](#credits)

<a name="introduction"></a>

## Introduction

**MaterialDialog** This Plugin will be useful to create simple, animated, and beautiful dialogs in your next Flutter app.
This library implements Airbnb's [_Lottie_](https://lottiefiles.com/) library to render After Effects animation in app.

<a name="types"></a>

## Types of Dialog

**MaterialDialog** library provides two types of dialog i.e.

<table style="width:100%">
  <tr>
    <th><b>1. Material Dialog</b></th>
    <th><b>2. Bottom Sheet Material Dialog</b></th>
  </tr>
  <tr>
    <td>A normal material dialog which can have one or two buttons.</td>
    <td>A Bottom Sheet material dialog which can have one or two buttons, is showed from bottom of device.</td> 
  </tr>
  <tr>
    <td align="center"><img src="https://github.com/Ezaldeen99/material_dialogs/raw/master/gifs/Screenshot_20201204_173336.png" width="75%"/></td>
    <td align="center"><img src="https://github.com/Ezaldeen99/material_dialogs/raw/master/gifs/Screenshot_20201204_173347.png" width="75%"/></td> 
  </tr>
</table>

<a name="implementation"></a>

## Implementation

Implementation of Material Dialog library is so easy. You can check [/example](/example) directory for demo. Let's have look talk in details about it.

<a name="install"></a>

### install

#### i. pubspec

In `pubspec.yaml`

```yaml
dependencies:
  material_dialogs: _latest_version
```

Now in your Dart code, you can use:

```dart
import 'package:material_dialogs/material_dialogs.dart';
```

Details see [pub.dev](https://pub.dev/packages/material_dialogs/install).

<a name="createDialog"></a>

### Create Dialog

As there are two types of dialogs in library. Material Dialogs are instantiated as follows.
<a name="createMaterialDialog"></a>

#### i. Material Dialog

`Dialogs` class will be used to create your dialog, below is an example to show your dialog in the app.

```dart
Dialogs.materialDialog(
          msg: 'Are you sure ? you can\'t undo this',
          title: "Delete",
          color: Colors.white,
          context: context,
          actions: [
            IconsOutlineButton(
              onPressed: () {},
              text: 'Cancel',
              iconData: Icons.cancel_outlined,
              textStyle: TextStyle(color: Colors.grey),
              iconColor: Colors.grey,
            ),
            IconsButton(
              onPressed: () {},
              text: 'Delete',
              iconData: Icons.delete,
              color: Colors.red,
              textStyle: TextStyle(color: Colors.white),
              iconColor: Colors.white,
            ),
          ])
```

<img align="center" src="https://github.com/Ezaldeen99/material_dialogs/raw/master/gifs/normal.gif" width="300"/>

`IconsOutlineButton` and `IconsButton` are both buttons widgets provided by the plugin to make things easier for you [read more](#icons_buttons), you can use any other buttons if you want.

<a name="createBsMaterialDialog"></a>

#### ii. Bottom Sheet Material Dialog

`Dialogs` class will be used to create your dialog, use `bottomMaterialDialog`. Below is an example to show your dialog in the app.

```dart
Dialogs.bottomMaterialDialog(
          msg: 'Are you sure? you can\'t undo this action',
          title: 'Delete',
          context: context,
          actions: [
            IconsOutlineButton(
              onPressed: () {},
              text: 'Cancel',
              iconData: Icons.cancel_outlined,
              textStyle: TextStyle(color: Colors.grey),
              iconColor: Colors.grey,
            ),
            IconsButton(
              onPressed: () {},
              text: 'Delete',
              iconData: Icons.delete,
              color: Colors.red,
              textStyle: TextStyle(color: Colors.white),
              iconColor: Colors.white,
            ),
          ]),
```

<img align="center" src="https://github.com/Ezaldeen99/material_dialogs/raw/master/gifs/bottom.gif" width="300"/>

<a name="showAnims"></a>

### Show Animations

Animations in this library are implemented using Lottie animation library. You can get free animations files [here](https://lottiefiles.com/).

`*.json` file downloaded from _LottieFiles_ should be placed in flutter project.

For example, here `cong_example.json` animation file is used in the `assets` folder to show congratulations animation in the example app.

In code, set `lottieBuilder` arg in Widget to set Animation to the dialog.

```dart
Dialogs.materialDialog(
          color: Colors.white,
          msg: 'Congratulations, you won 500 points',
          title: 'Congratulations',
          lottieBuilder: Lottie.asset(
          'assets/cong_example.json',
          fit: BoxFit.contain,
          ),
          context: context,
          actions: [
            IconsButton(
              onPressed: () {},
              text: 'Claim',
              iconData: Icons.done,
              color: Colors.blue,
              textStyle: TextStyle(color: Colors.white),
              iconColor: Colors.white,
            ),
          ]),
```

<a name="icons_buttons"></a>

## Icons buttons

The plugin provide you some out of the box customized buttons to help you creating your dialog.

#### IconsOutlineButton

This widget helps you create an outline button easily

```dart
    IconsOutlineButton(
      onPressed: () {},
      text: 'Cancel',
      iconData: Icons.cancel_outlined,
      textStyle: TextStyle(color: Colors.grey),
      iconColor: Colors.grey,
    ),
```

#### IconsButton

This widget helps you create a material button with icons in few lines of code

```dart
    IconsButton(
      onPressed: () {},
      text: 'Delete',
      iconData: Icons.delete,
      color: Colors.red,
      textStyle: TextStyle(color: Colors.white),
      iconColor: Colors.white,
    ),
```

<img align="center" src="https://github.com/Ezaldeen99/material_dialogs/raw/master/gifs/single_btn.png" width="300"/>

#### CustomView

You can add your own Widget inside the dialog by using the customView attribute and CustomViewPosition to place it wherever you want.

| Possible values        |
| ---------------------- |
| BEFORE_ANIMATION       |
| BEFORE_TITLE (Default) |
| BEFORE_MESSAGE         |
| BEFORE_ACTION          |
| AFTER_ACTION           |

```dart
Dialogs.materialDialog(
          color: Colors.white,
          msg: 'Congratulations, you won 500 points',
          title: 'Congratulations',
          lottieBuilder: Lottie.asset(
          'assets/cong_example.json',
          fit: BoxFit.contain,
          ),
          customView: MySuperWidget(),
          customViewPosition: CustomViewPosition.BEFORE_ACTION,
          context: context,
          actions: [
            IconsButton(
              onPressed: () {},
              text: 'Claim',
              iconData: Icons.done,
              color: Colors.blue,
              textStyle: TextStyle(color: Colors.white),
              iconColor: Colors.white,
            ),
          ]),
```

# Limitations

it's better to make your animation to have the same background color as your dialog's background color, please use [lottie editor](https://lottiefiles.com/editor) to remove the background layer of your animation or make it same as your dialog background color before using it in the plugin.

<a name="contribute"></a>

## Contribute

Let's develop with collaborations. We would love to have contributions by raising issues and opening PRs. Filing an issue before PR is must.

<a name="credits"></a>

## Credits

This library is built using following open-source libraries.

- [Lottie for Flutter](https://pub.dev/packages/lottie)
- [MaterialDialog-Android](https://github.com/PatilShreyas/MaterialDialog-Android) for inspiration

## License

Project is published under the Apache 2.0 license. Feel free to clone and modify repo as you want, but don't forget to add reference to authors :)


CHANGELOG.md
## [1.1.4]

- chore: update dependencies

## [1.1.3]

- adding `dialogWidth` in order to support flexible dialogue width

## [1.1.2]

- allow returning values in on close
- update dependencies

## [1.1.1]

- fix bugs

## [1.1.0]

# Breakout changes:

- remove `animation`, `frameRate`, `reverse` and replace it with `lottieBuilder`

# New features

- add on close callback
- adding custom view, in case you wanted to use only image or different size animations.

## [1.0.0]

- Migrate to Sound null safety
- remove deprecated code

## [0.1.5]

- Update dependencies

## [0.1.4]

- Adding Custom widget in case anyone needs to replace the animation
  with a custom view
- Update dependencies

## [0.1.3]

- Adding barrier dismissible

## [0.1.2]

- Fixing bottom dialog null frame rate bug

## [0.1.1]

- Fixing some bugs

## [0.1.0]

- Adding animationFrameRate
- Adding animationRepeat
- Adding animationAnimate
- Adding animationReverse

## [0.0.6+1]

-allow custom padding on both IconsButton and OutlinedIconsButton

## [0.0.6]

-mark title and msg variables as not required

## [0.0.5]

- allow only description dialogs
- allow only titles dialogs
- allow only buttons dialogs

## [0.0.4]

- adding RTL layouts support

## [0.0.3+1]

- edit examples

## [0.0.3]

- breakout changes
- use actions widgets list to add buttons.
- add out of the box customized buttons widgets.

## [0.0.2]

- fix single button padding bug and edit ReadMe.

## [0.0.1]

- first release.


example/lib/main.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:google_fonts/google_fonts.dart';
import 'package:lottie/lottie.dart';
import 'package:material_dialogs/material_dialogs.dart';
import 'package:material_dialogs/widgets/buttons/icon_button.dart';
import 'package:material_dialogs/widgets/buttons/icon_outline_button.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Material design Dialogs',
        theme: ThemeData(
          textTheme: GoogleFonts.montserratTextTheme(
            Theme.of(context).textTheme,
          ),
          primarySwatch: Colors.blue,
        ),
        home: SafeArea(
          child: Scaffold(
              backgroundColor: Colors.white,
              appBar: AppBar(
                title: Text("Material design Dialogs"),
              ),
              body: TestPage()),
        ));
  }
}

class TestPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return TestState();
  }
}

class TestState extends State<TestPage> {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.center,
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          btn1(context),
          btn2(context),
          btn3(context),
          btn4(context),
        ],
      ),
    );
  }

  Widget btn1(BuildContext context) {
    return MaterialButton(
      color: Colors.grey[300],
      minWidth: 300,
      onPressed: () => Dialogs.materialDialog(
          msg: 'Are you sure ? you can\'t undo this',
          title: "Delete",
          color: Colors.white,
          context: context,
          dialogWidth: kIsWeb ? 0.3 : null,
          onClose: (value) => print("returned value is '$value'"),
          actions: [
            IconsOutlineButton(
              onPressed: () {
                Navigator.of(context).pop(['Test', 'List']);
              },
              text: 'Cancel',
              iconData: Icons.cancel_outlined,
              textStyle: TextStyle(color: Colors.grey),
              iconColor: Colors.grey,
            ),
            IconsButton(
              onPressed: () {},
              text: "Delete",
              iconData: Icons.delete,
              color: Colors.red,
              textStyle: TextStyle(color: Colors.white),
              iconColor: Colors.white,
            ),
          ]),
      child: Text("Show Material Dialog"),
    );
  }

  Widget btn2(BuildContext context) {
    return MaterialButton(
      minWidth: 300,
      color: Colors.grey[300],
      onPressed: () => Dialogs.bottomMaterialDialog(
          msg: 'Are you sure? you can\'t undo this action',
          title: 'Delete',
          context: context,
          actions: [
            IconsOutlineButton(
              onPressed: () {
                Navigator.of(context).pop();
              },
              text: 'Cancel',
              iconData: Icons.cancel_outlined,
              textStyle: TextStyle(color: Colors.grey),
              iconColor: Colors.grey,
            ),
            IconsButton(
              onPressed: () {},
              text: 'Delete',
              iconData: Icons.delete,
              color: Colors.red,
              textStyle: TextStyle(color: Colors.white),
              iconColor: Colors.white,
            ),
          ]),
      child: Text("Show Bottom Material Dialog"),
    );
  }

  Widget btn3(BuildContext context) {
    return MaterialButton(
      minWidth: 300,
      color: Colors.grey[300],
      onPressed: () => Dialogs.materialDialog(
        color: Colors.white,
        msg: 'Congratulations, you won 500 points',
        title: 'Congratulations',
        lottieBuilder: Lottie.asset(
          'assets/cong_example.json',
          fit: BoxFit.contain,
        ),
        dialogWidth: kIsWeb ? 0.3 : null,
        context: context,
        actions: [
          IconsButton(
            onPressed: () {
              Navigator.of(context).pop();
            },
            text: 'Claim',
            iconData: Icons.done,
            color: Colors.blue,
            textStyle: TextStyle(color: Colors.white),
            iconColor: Colors.white,
          ),
        ],
      ),
      child: Text("Show animations Material Dialog"),
    );
  }

  Widget btn4(BuildContext context) {
    return MaterialButton(
      color: Colors.grey[300],
      minWidth: 300,
      onPressed: () => Dialogs.bottomMaterialDialog(
        msg: 'Congratulations, you won 500 points',
        title: 'Congratulations',
        color: Colors.white,
        lottieBuilder: Lottie.asset(
          'assets/cong_example.json',
          fit: BoxFit.contain,
        ),
        context: context,
        actions: [
          IconsButton(
            onPressed: () {
              Navigator.of(context).pop();
            },
            text: 'Claim',
            iconData: Icons.done,
            color: Colors.blue,
            textStyle: TextStyle(color: Colors.white),
            iconColor: Colors.white,
          ),
        ],
      ),
      child: Text("Show animations Bottom Dialog"),
    );
  }
}
